sys/net/gnrc/sock/tcp/gnrc_sock_tcp.c

```diff
@@ -0,0 +1,216 @@
+                    sock_tcp_t *queue_array, unsigned queue_len, uint16_t flags)
+{
+    /* Asserts defined by API. */
+    assert(queue != NULL);
+    assert(local != NULL);
+    assert(local->port != 0);
+    assert(queue_array != NULL);
+    assert(queue_len != 0);
+
+    /* Asserts to protect GNRC_TCP. Flags are not supported. */
+    assert(flags == 0);
+    assert(sizeof(sock_tcp_queue_t) == sizeof(gnrc_tcp_tcb_queue_t));
+    assert(sizeof(sock_tcp_t) == sizeof(gnrc_tcp_tcb_t));
+    assert(sizeof(sock_tcp_ep_t) == sizeof(gnrc_tcp_ep_t));
+
+    /* NOTE: GNRC_TCP and GNRC_SOCK_TCP types must have the same memory representation. */
+    gnrc_tcp_tcb_queue_t *tcb_queue = (gnrc_tcp_tcb_queue_t *) queue;
+    gnrc_tcp_tcb_t *tcbs = (gnrc_tcp_tcb_t *) queue_array;
+    gnrc_tcp_ep_t *ep = (gnrc_tcp_ep_t *) local;
+
+    /* Initialize given TCB queue, all given tcbs and forward call */
+    gnrc_tcp_tcb_queue_init(tcb_queue);
+    for (unsigned i = 0; i < queue_len; ++i) {
+        gnrc_tcp_tcb_init(&tcbs[i]);
+    }
+    return gnrc_tcp_listen(tcb_queue, tcbs, queue_len, ep);
+}
+
+void sock_tcp_disconnect(sock_tcp_t *sock)
+{
+    /* Asserts defined by API. */
+    assert(sock != NULL);
+
+    /* Asserts to protect GNRC_TCP. */
+    assert(sizeof(sock_tcp_t) == sizeof(gnrc_tcp_tcb_t));
+
+    /* NOTE: GNRC_TCP and GNRC_SOCK_TCP types must have the same memory representation. */
+    gnrc_tcp_tcb_t *tcb = (gnrc_tcp_tcb_t *) sock;
+    gnrc_tcp_close(tcb);
+}
+
+void sock_tcp_stop_listen(sock_tcp_queue_t *queue)
+{
+    /* Asserts defined by API. */
+    assert(queue != NULL);
+
+    /* Asserts to protect GNRC_TCP. */
+    assert(sizeof(sock_tcp_queue_t) == sizeof(gnrc_tcp_tcb_queue_t));
+
+    /* NOTE: GNRC_TCP and GNRC_SOCK_TCP types must have the same memory representation. */
+    gnrc_tcp_tcb_queue_t *tcb_queue = (gnrc_tcp_tcb_queue_t *) queue;
+    gnrc_tcp_stop_listen(tcb_queue);
+}
+
+int sock_tcp_get_local(sock_tcp_t *sock, sock_tcp_ep_t *ep)
+{
+    /* Asserts defined by API. */
+    assert(sock != NULL);
+    assert(ep != NULL);
+
+    /* Asserts to protect GNRC_TCP. */
+    assert(sizeof(sock_tcp_t) == sizeof(gnrc_tcp_tcb_t));
+    assert(sizeof(sock_tcp_ep_t) == sizeof(gnrc_tcp_ep_t));
+
+    /* NOTE: GNRC_TCP and GNRC_SOCK_TCP types must have the same memory representation. */
+    gnrc_tcp_tcb_t *tcb = (gnrc_tcp_tcb_t *) sock;
+    gnrc_tcp_ep_t *local = (gnrc_tcp_ep_t *) ep;
+    return gnrc_tcp_get_local(tcb, local);
+}
+
+int sock_tcp_get_remote(sock_tcp_t *sock, sock_tcp_ep_t *ep)
+{
+    /* Asserts defined by API. */
+    assert(sock != NULL);
+    assert(ep != NULL);
+
+    /* Asserts to protect GNRC_TCP. */
+    assert(sizeof(sock_tcp_t) == sizeof(gnrc_tcp_tcb_t));
+    assert(sizeof(sock_tcp_ep_t) == sizeof(gnrc_tcp_ep_t));
+
+    /* NOTE: GNRC_TCP and GNRC_SOCK_TCP types must have the same memory representation. */
+    gnrc_tcp_tcb_t *tcb = (gnrc_tcp_tcb_t *) sock;
+    gnrc_tcp_ep_t *remote = (gnrc_tcp_ep_t *) ep;
```

Comment:
The Idea is to keep the layers seperated. sock is an unified interface allowing multiple implementations network protocol implementations to coexist depending on the build configuration. gnrc on the other hand is a specific network stack that should have no dependency to sock. 

=> Its fine for sock depend on gnrc but gnrc should be independent of sock.

Reply:
I don't see how implicitly relying that those two structs are the same is any better than including a generic header file :fearful:

Reply:
I'll take a look at your suggestion.

Reply:
Infact I do not include any sock header within gnrc_tcp (at least my grep skills say so).

How about "typedef struct _sock_tl_ep gnrc_tcp_ep_t;" within gnrc_tcp?

Reply:
In `sys/net/gnrc/sock/tcp/gnrc_sock_tcp.c` you include `net/sock/tcp.h` which in turn includes `net/sock.h`

Reply:
gnrc_sock_tcp is the gnrc_tcp adapter for sock_tcp. Thats something different that gnrc_tcp.

Anyway I'll use the sock type in gnrc_tcp because its the better solution ^^

Reply:
IMHO `gnrc_tcp` should still be usable without `sock_tcp`, so having a dedicated end-point type for it is the right way to go (even if it is a verbatim copy).

Reply:
To prevent casting, we could have a union inside `gnrc_tcp_ep_t` that maps it to `sock_tcp_ep_t`.

Reply:
`net/sock.h` lives in `sys/include`, so it can always be included - no `sock_tcp` module needed

Reply:
But `gnrc_sock_tcp` might not be compiled, leading to a missing `sock_types.h`.

Reply:
Nevertheless, I think it is confusing to use a type of a layer that is not there. It also makes `gnrc_tcp` dependent on the `sock` module (even if only by header) which is a bad modularity design IMHO.

Reply:
I think I found a solution that I solves the Issue "as clean as possible".
See: https://github.com/RIOT-OS/RIOT/pull/16494/commits/b1d0cef158ccdeae1841a8c0d9b240c3a7771860

Any thoughts on this?

Reply:
That's basically a variant on what I proposed :-).

Reply:
Just without a union